api/spec/api_docs/definitions/resource.rb

```diff
@@ -1,78 +1,146 @@
+      end
+
+      def create_request
+        definition = make_request(__callee__, create_attributes || request_attributes)
+        definition = Definitions::DryTypesToJson.convert(definition)
+        transform_keys(definition)
+      end
+
+      def update_attributes
+        nil
+      end
+
+      def create_attributes
+        nil
+      end
+
+      def request_attributes
+        ::Types::Hash.schema(self::ATTRIBUTES.merge(attributes).except(*self::READ_ONLY))
+      end
+
+      def request_relationships
+        Type.object(properties: {})
+      end
+
+      def required_create_attributes
+        const_defined?(:REQUIRED_CREATE_ATTRIBUTES) ? self::REQUIRED_CREATE_ATTRIBUTES : required_attributes
+      end
+
+      def required_update_attributes
+        const_defined?(:REQUIRED_UPDATE_ATTRIBUTES) ? self::REQUIRED_UPDATE_ATTRIBUTES : required_attributes
+      end
+
+      def required_attributes
+        const_defined?(:REQUIRED_ATTRIBUTES) ? self::REQUIRED_ATTRIBUTES : []
+      end
+
+      ######################################
+      ############# RESPONSES ##############
+      ######################################
+
+      def resource_data(attr, realtion)
+        ::Types::Hash.schema(
+          id: ::Types::Serializer::ID,
+          type: ::Types::String.meta(example: type),
+          attributes: attr,
+          relationships: realtion,
+          meta: ::Types::Serializer::Meta
+        )
+      end
+
+      def resource_response_data
+        return collection_response_data if serializer.partial_only?
+
+        resource_data(
+          response_attributes(full_attributes),
+          response_relationships(full_relationships)
         )
+      end
+
+      def resource_response
+        definition = ::Types::Hash.schema(data: resource_response_data)
+
+        definition = Definitions::DryTypesToJson.convert(definition)
         debug(__callee__, definition)
         transform_keys(definition)
       end
 
       def collection_response
-        definition = Type.object(
-          properties: {
-            data: Type.array(
-              items: collection_resource_response || resource_response
-            )
-          }
+        definition = ::Types::Hash.schema(
+          data: ::Types::Array.of(collection_response_data)
         )
+
+        definition = Definitions::DryTypesToJson.convert(definition)
         debug(__callee__, definition)
         transform_keys(definition)
       end
 
-      def update_request
-        definition = make_request(__callee__, update_attributes || request_attributes)
-        transform_keys(definition)
+      def collection_response_data
+        resource_data(
+          response_attributes(attributes),
+          response_relationships(relationships)
+        )
       end
 
-      def create_request
-        definition = make_request(__callee__, create_attributes || request_attributes)
-        transform_keys(definition)
+      def response_relationships(realtion)
```

Comment:
Oh, I see the confusion. I was changing the name from `relationships` to `relation` not because the input changed, but because I wanted to make the distinction between the variable `relationships` and the method `relationships`. While I understand that ruby's scoping ensures that in the method it will use the local variable `relationships`, I felt it made it more confusing than it needed to be. 

I do now see with your comment how this just switched one confusion for another. Do you have any recommendations for situations like that?

Reply:
I'd like you to solve this problem without my input—given my responsibilities at CIC, I don't have the bandwidth to solve problems that are this granular. I need you to be able to solve this. I can give you some guidance, however. Method names should clearly describe what the method is doing. Your explanation just doesn't make sense—if we have a method called "relationships", I would expect it to return a collection of relationships. If that method is changed to "relation," I would expect it to return a single relation. It's weird to change it to "relation" since we universally refer to these things as relationships throughout the code, right? It's weird to name a method that returns multiple things in a way that implies it only returns one thing. It's weird to change methods on a class to something confusing because you don't want a local variable to overlap. Please go through this and ensure that your changes to not make it harder to understand what's happening in this class.